Segfaults in header reading code under certain setups

It's not clear exact what factors cause this:

config:
```
  [config]
  Version = 1.0.0
  SnapshotType = LGadget-2
  HaloType = nil
  TreeType = nil
  SnapshotFormat =
/vol/porciani2/data1/emiliord/accretion/data/halo1202_4096_baryons/nbody/gdgt_%03d_l1
  SnapshotFormatMeanings = Snapshot
  SnapMin = 681
  SnapMax = 681
  MemoDir = /vol/porciani2/data1/egaraldi/accretion/shellfish/memodir
  Endianness = SystemOrder
  ValidateFormats = false
  Threads = -1
  Logging = nil
```

Stacktrace:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x4cc989]

goroutine 1 [running]:
github.com/phil-mansfield/shellfish/cmd/memo.readUnmemoizedHeaders(0x2a9,
0x0, 0x0, 0xc420126150, 0x8e0240, 0xc4200ccf30, 0xe, 0x5f8660, 0x4,
0x25, ...)

/vol/aibn230/data1/egaraldi/golang/go/src/github.com/phil-mansfield/shellfish/cmd/memo/memoize.go:250
+0x169
github.com/phil-mansfield/shellfish/cmd/memo.ReadHeaders(0x2a9, 0x0,
0x0, 0xc420126150, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)

/vol/aibn230/data1/egaraldi/golang/go/src/github.com/phil-mansfield/shellfish/cmd/memo/memoize.go:270
+0x252
github.com/phil-mansfield/shellfish/cmd.loop(0xc4201505f8, 0x1, 0x1,
0xc420150610, 0x1, 0x1, 0xc4200c6a80, 0x4, 0x4, 0xc4200d8080, ...)

/vol/aibn230/data1/egaraldi/golang/go/src/github.com/phil-mansfield/shellfish/cmd/shell.go:303
+0x955
github.com/phil-mansfield/shellfish/cmd.(*ShellConfig).Run(0xc4200d8080,
0xc4201281c0, 0xc420126150, 0xc4200dc1c0, 0x1, 0x2, 0x0, 0x0, 0x5, 0x1, ...)

/vol/aibn230/data1/egaraldi/golang/go/src/github.com/phil-mansfield/shellfish/cmd/shell.go:239
+0x49d
main.main()

/vol/aibn230/data1/egaraldi/golang/go/src/github.com/phil-mansfield/shellfish/shellfish.go:407
+0x51c
```

--- a/cmd/cmd.go
+++ b/cmd/cmd.go
@@ -65,6 +65,8 @@ type GlobalConfig struct {
 	Threads         int64
 
 	Logging string
+
+	GadgetNpartNum int64
 }
 
 var _ Mode = &GlobalConfig{}
@@ -102,6 +104,7 @@ func (config *GlobalConfig) ReadConfig(fname string, flags []string) error {
 
 	vars.Int(&config.Threads, "Threads", -1)
 	vars.String(&config.Logging, "Logging", "nil")
+	vars.Int(&config.GadgetNpartNum, "GadgetNpartNum", 2)
 
 	if err := parse.ReadConfig(fname, vars); err != nil {
 		return err
@@ -276,6 +279,14 @@ func (config *GlobalConfig) validate() error {
 			len(config.HaloValueNames), len(config.HaloValueComments),
 		)
 	}
+
+	if config.GadgetNpartNum > 2 || config.GadgetNpartNum <= 0 {
+		return fmt.Errorf(
+			"GadgetNpartNum set to %d, but the only valid values are 1 and 2.",
+			config.GadgetNpartNum,
+		)
+	}
+
 	
 	return validateFormat(config)
 }
@@ -505,7 +516,18 @@ Threads = -1
 # nil - no logging is performed.
 # performance - runtime and memory consumption logging are written to stderr.
 # debugging - debugging information is written to stderr
-Logging = nil`, version.SourceVersion)
+Logging = nil
+
+# GadgetNpartNum is an optional variable which should only be set when using
+# Gadget files. If your Gadget files use two elements of Npart and Nall to
+# represent the number of dark matter particles in your simulation, set this
+# variable to 2. If every element corresponds to a different particle species,
+# set this variable to 1. GadgetNpartNum defaults to the most common value, 2.
+#
+# If you don't know what your Gadget file does, leave this alone: Shellfish will
+# fail and tell you to change this variable.
+# GadgetNpartNum = 2
+`, version.SourceVersion)
 }
 
 // Run is a dummy method which allows GlobalConfig to conform to the Mode

--- a/cmd/coord.go
+++ b/cmd/coord.go
@@ -94,6 +94,7 @@ func (config *CoordConfig) Run(
 	buf, err := getVectorBuffer(
 		e.ParticleCatalog(snaps[0], 0),
 		gConfig.SnapshotType, gConfig.Endianness,
+		gConfig.GadgetNpartNum,
 	)
 	if err != nil {
 		return nil, err

--- a/cmd/id.go
+++ b/cmd/id.go
@@ -240,6 +240,7 @@ func (config *IDConfig) Run(
 		buf, err = getVectorBuffer(
 			e.ParticleCatalog(snaps[0], 0),
 			gConfig.SnapshotType, gConfig.Endianness,
+			gConfig.GadgetNpartNum,
 		)
 		if err != nil {
 			return nil, err
@@ -254,6 +255,7 @@ func (config *IDConfig) Run(
 		buf, err = getVectorBuffer(
 			e.ParticleCatalog(snaps[0], 0),
 			gConfig.SnapshotType, gConfig.Endianness,
+			gConfig.GadgetNpartNum,
 		)
 		if err != nil {
 			return nil, err
@@ -337,6 +339,7 @@ func getMassIDRange(
 	buf, err := getVectorBuffer(
 		e.ParticleCatalog(int(config.snap), 0),
 		gConfig.SnapshotType, gConfig.Endianness,
+		gConfig.GadgetNpartNum,
 	)
 
 	if err != nil { return err }

--- a/cmd/memo/memoize.go
+++ b/cmd/memo/memoize.go
@@ -242,11 +242,14 @@ func (f IntFinder) Find(rid int) (int, bool) {
 func readUnmemoizedHeaders(
 	snap int, buf io.VectorBuffer, e *env.Environment,
 ) ([]io.Header, []string, error) {
+	fmt.Println("Buffer:", buf)
 	files := make([]string, e.Blocks())
 	hds := make([]io.Header, e.Blocks())
 
 	for i := range files {
+		fmt.Println("Block:", i)
 		files[i] = e.ParticleCatalog(snap, i)
+		fmt.Println("File:", files[i])
 		err := buf.ReadHeader(files[i], &hds[i])
 		if err != nil {
 			return nil, nil, err

--- a/cmd/memo/memoize.go
+++ b/cmd/memo/memoize.go
@@ -243,6 +243,7 @@ func readUnmemoizedHeaders(
 	snap int, buf io.VectorBuffer, e *env.Environment,
 ) ([]io.Header, []string, error) {
 	fmt.Println("Buffer:", buf)
+	fmt.Println("BlockNum:", e.Blocks())
 	files := make([]string, e.Blocks())
 	hds := make([]io.Header, e.Blocks())
 

--- a/cmd/prof.go
+++ b/cmd/prof.go
@@ -284,6 +284,7 @@ func (config *ProfConfig) Run(
 	buf, err := getVectorBuffer(
 		e.ParticleCatalog(snaps[0], 0),
 		gConfig.SnapshotType, gConfig.Endianness,
+		gConfig.GadgetNpartNum,
 	)
 	if err != nil {
 		return nil, err

--- a/cmd/shell.go
+++ b/cmd/shell.go
@@ -233,6 +233,7 @@ func (config *ShellConfig) Run(
 	buf, err := getVectorBuffer(
 		e.ParticleCatalog(snaps[0], 0),
 		gConfig.SnapshotType, gConfig.Endianness,
+		gConfig.GadgetNpartNum,
 	)
 	if err != nil {
 		return nil, err

--- a/cmd/shell.go
+++ b/cmd/shell.go
@@ -234,6 +234,9 @@ func (config *ShellConfig) Run(
 		e.ParticleCatalog(snaps[0], 0),
 		gConfig.SnapshotType, gConfig.Endianness,
 	)
+	if err != nil {
+		return nil, err
+	}
 
 	err = loop(ids, snaps, coords, config, buf, e, out, gConfig.Threads)
 	if err != nil {

--- a/cmd/stats.go
+++ b/cmd/stats.go
@@ -217,6 +217,7 @@ func (config *StatsConfig) Run(
 	buf, err := getVectorBuffer(
 		e.ParticleCatalog(snaps[0], 0),
 		gConfig.SnapshotType, gConfig.Endianness,
+		gConfig.GadgetNpartNum,
 	)
 	if err != nil {
 		return nil, err

--- a/cmd/utils.go
+++ b/cmd/utils.go
@@ -6,13 +6,13 @@ import (
 )
 
 func getVectorBuffer(
-	fname, typeString, endiannessString string,
+	fname, typeString, endiannessString string, npartNum int64,
 ) (io.VectorBuffer, error) {
 	switch typeString {
 	case "gotetra":
 		return io.NewGotetraBuffer(fname)
 	case "LGadget-2":
-		return io.NewLGadget2Buffer(fname, endiannessString)
+		return io.NewLGadget2Buffer(fname, endiannessString, npartNum)
 	case "ARTIO":
 		return io.NewARTIOBuffer(fname)
 	}

--- a/io/lgadget2.go
+++ b/io/lgadget2.go
@@ -22,10 +22,23 @@ type lGadget2Header struct {
 	Padding [88]byte
 }
 
-func (gh *lGadget2Header) postprocess(xs [][3]float32, out *Header) {
+func (gh *lGadget2Header) postprocess(
+	xs [][3]float32, npartNum int64, out *Header,
+) {
 	// Assumes the catalog has already been checked for corruption.
+	if npartNum == 2 {
+		if gh.NPart[0] > 100 * 1000 {
+			panic(
+				"Simulation contains too many particles. This is probably " +
+				"because GadgetNpartNum is set to 2 when it " +
+				"should be set to 1.",
+			)
+		}
+		out.N = int64(gh.NPart[1]) + int64(gh.NPart[0]) << 32
+	} else {
+		out.N = int64(gh.NPart[0])
+	}
 
-	out.N = int64(gh.NPart[1]) + int64(gh.NPart[0])<<32
 	out.TotalWidth = gh.BoxSize
 
 	out.Cosmo.Z = gh.Redshift
@@ -78,7 +91,21 @@ func (buf *LGadget2Buffer) readLGadget2Particles(
 	_ = readInt32(f, order)
 	binary.Read(f, binary.LittleEndian, gh)
 	_ = readInt32(f, order)
-	count := int(int64(gh.NPart[1]) + int64(gh.NPart[0])<<32)
+
+	var count int
+	if buf.npartNum == 2 {
+		if gh.NPart[0] > 100 * 1000 {
+			return nil, nil, nil, nil, fmt.Errorf(
+				"Simulation contains too many particles. This is probably " +
+				"because GadgetNpartNum is set to 2 when it " +
+				"should be set to 1.",
+			)
+		}
+		count = int(int64(gh.NPart[1]) + int64(gh.NPart[0]) << 32)
+	} else {
+		count = int(gh.NPart[0])
+	}
+
 	xsBuf = expandVectors(xsBuf[:0], count)
 	vsBuf = expandVectors(vsBuf[:0], count)
 	idsBuf = expandInts(idsBuf[:0], count)
@@ -163,16 +190,19 @@ func expandInts(ints []int64, n int) []int64 {
 }
 
 type LGadget2Buffer struct {
-	open   bool
-	order  binary.ByteOrder
-	hd     lGadget2Header
-	mass   float32
-	xs, vs [][3]float32
-	ms     []float32
-	ids    []int64
+	open     bool
+	order    binary.ByteOrder
+	hd       lGadget2Header
+	mass     float32
+	xs, vs   [][3]float32
+	ms       []float32
+	ids      []int64
+	npartNum int64
 }
 
-func NewLGadget2Buffer(path, orderFlag string) (VectorBuffer, error) {
+func NewLGadget2Buffer(
+	path, orderFlag string, npartNum int64,
+) (VectorBuffer, error) {
 	var order binary.ByteOrder = binary.LittleEndian
 	switch orderFlag {
 	case "LittleEndian":
@@ -194,8 +224,21 @@ func NewLGadget2Buffer(path, orderFlag string) (VectorBuffer, error) {
 		Z: buf.hd.Redshift, OmegaM: buf.hd.Omega0,
 		OmegaL: buf.hd.OmegaLambda, H100: buf.hd.HubbleParam,
 	}
-	totCount := int64(buf.hd.NPartTotal[1]) + int64(buf.hd.NPartTotal[0])<<32
+	var totCount int64
+	if buf.npartNum == 2 {
+		if buf.hd.NPartTotal[0] > 100 * 1000 {
+			return nil, fmt.Errorf(
+				"Simulation contains too many particles. This is probably " +
+				"because GadgetNpartNum is set to 2 when it " +
+				"should be set to 1.",
+			)
+		}
+		totCount = int64(buf.hd.NPartTotal[1])+int64(buf.hd.NPartTotal[0])<<32
+	} else {
+		totCount = int64(buf.hd.NPartTotal[0])
+	}
 	buf.mass = calcUniformMass(totCount, buf.hd.BoxSize, c)
+	buf.npartNum = npartNum
 
 	return buf, nil
 }
@@ -237,7 +280,7 @@ func (buf *LGadget2Buffer) ReadHeader(fname string, out *Header) error {
 		return err
 	}
 
-	buf.hd.postprocess(xs, out)
+	buf.hd.postprocess(xs, buf.npartNum, out)
 
 	return nil
 }

--- a/io/lgadget2.go
+++ b/io/lgadget2.go
@@ -34,11 +34,11 @@ func (gh *lGadget2Header) postprocess(
 				"should be set to 1.",
 			)
 		}
-		out.N = int64(gh.NPart[1]) + int64(gh.NPart[0]) << 32
+		out.N = int64(gh.NPart[1]) + int64(uint32(gh.NPart[0])) << 32
 	} else {
 		out.N = int64(gh.NPart[0])
 	}
-
+	
 	out.TotalWidth = gh.BoxSize
 
 	out.Cosmo.Z = gh.Redshift
@@ -101,7 +101,7 @@ func (buf *LGadget2Buffer) readLGadget2Particles(
 				"should be set to 1.",
 			)
 		}
-		count = int(int64(gh.NPart[1]) + int64(gh.NPart[0]) << 32)
+		count = int(int64(gh.NPart[1]) + int64(uint32(gh.NPart[0])) << 32)
 	} else {
 		count = int(gh.NPart[0])
 	}
@@ -203,6 +203,7 @@ type LGadget2Buffer struct {
 func NewLGadget2Buffer(
 	path, orderFlag string, npartNum int64,
 ) (VectorBuffer, error) {
+	
 	var order binary.ByteOrder = binary.LittleEndian
 	switch orderFlag {
 	case "LittleEndian":
@@ -224,6 +225,9 @@ func NewLGadget2Buffer(
 		Z: buf.hd.Redshift, OmegaM: buf.hd.Omega0,
 		OmegaL: buf.hd.OmegaLambda, H100: buf.hd.HubbleParam,
 	}
+	
+	buf.npartNum = npartNum
+	
 	var totCount int64
 	if buf.npartNum == 2 {
 		if buf.hd.NPartTotal[0] > 100 * 1000 {
@@ -233,12 +237,13 @@ func NewLGadget2Buffer(
 				"should be set to 1.",
 			)
 		}
-		totCount = int64(buf.hd.NPartTotal[1])+int64(buf.hd.NPartTotal[0])<<32
+		totCount = int64(buf.hd.NPartTotal[1]) + 
+			int64(uint32(buf.hd.NPartTotal[0]))<<32
 	} else {
 		totCount = int64(buf.hd.NPartTotal[0])
 	}
+	
 	buf.mass = calcUniformMass(totCount, buf.hd.BoxSize, c)
-	buf.npartNum = npartNum
 
 	return buf, nil
 }

